py-rse/style.Rmd

```diff
@@ -4,207 +4,255 @@
-    comment on the first two or three and simply put a check beside the others—the reader
-    won't need the comment repeated.
+```python
+substrate[i, j]+overlay[i, j]
+```
 
-Don't ask people to do extra work.
-:   Nobody enjoys fixing bugs and style violations.
-    Asking them to add a few features while they're at it is rude.
+**Put two blank links between each function definition.**
 
-Don't let people break these rules just because they're frequent contributors or in positions of power.
-:   The culture of any organization is shaped by the worst behavior it is willing to tolerate @Grue2015.
-    If you let people be rude to one another,
-    *that* is your culture.
+This helps the eye see where one ends and the next begins,
+though the fact that functions always start in the first column helps as well.
 
-Be specific in replies to reviewers.
-:   If someone has suggested a better variable name,
-    you can probably simply fix it.
-    If someone has suggested a major overhaul to an algorithm,
-    you should reply to their comment to point at the commit that includes the fix.
+### Naming
 
-Thank your reviewers.
-:   If someone has taken the time to read your code carefully,
-    thank them for doing it.
+**Use `ALL_CAPS_WITH_UNDERSCORES` for global variables.**
 
-## What are the standard rules of good style for Python programs? {#py-rse-style-pep8}
+This convention is inherited from C,
+which was used to write the first version of Python.
+In that language,
+it was used to indicate a constant whose value couldn't be modified;
+Python doesn't enforce that rule,
+but `SHOUTING_AT_PROGRAMMERS` helps remind them that
+some things shouldn't be messed with.
 
-The single most important rule of style is to be consistent,
-both internally and with other programs.
-Python's standard style is called [PEP-8][pep-8];
-the term "PEP" is short for "Python Enhancement Proposal",
-and PEP-8 has become the most widely used standard for Python coding.
-Some of its rules are listed below,
-along with others borrowed from "[Code Smells and Feels][code-smells-and-feels]":
+**Use `lower_case_with_underscores` for the names of functions and variables.**
 
-FIXME: add before-and-after examples for each of these.
+Research on naming conventions has produced mixed results @Bink2012,@Scha2018,
+but Python has (mostly) settled on underscored names for most things.
+This naming convention is called [snake case][snake-case] or [pothole case][pothole-case];
+you should only use [`CamelCase`][camel-case] for classes,
+which are outside the scope of this lesson.
 
-FIXME: move this to the novice material.
+**Avoid abbreviations in function and variable names.**
 
-### Always indent code blocks using 4 spaces, and use spaces instead of tabs.
+They can be ambiguous,
+and can be be hard for non-native speakers to understand.
+This rule doesn't necessarily you will have to do more typing:
+a good programming editor will [auto-complete][auto-completion] names for you.
```

Comment:
Hmmm, could we also have a note that this is not an OK signal to use giant variable names, that is really hard to read as well.

Reply:
Is there anything in PEP-8 about overly-long names?

Reply:
Nope, my comment was about readability in general. To be clear, I am not saying that abbreviations should be used when a variable name exceeds a certain length, but that if you can think of a shorter name that is as descriptive, then use that instead. It should be clear and concise, and use commonly accepted abbreviations (e.g. "id" instead of "identification_number").

Reply:
Again, PEP-8 doesn't make a recommendation, and I would personally prefer `identification_number` over `id` for a variable with wide scope...

Reply:
To me, this paragraph is about general best practices in coding, not limited to those mentioned in PEP 8 (note that PEP 8  does not recommend against using abbreviations in names either).

I agree that verbosity depends on scope, but there is also meaningless verbosity that doesn't convey more information, which is where I think `id` falls (another example would be expanding `usa` to `united_states_of_america`, and [this HN comment](https://news.ycombinator.com/item?id=12311031) illustrates and example not related to abbreviations). This [SO answer](https://stackoverflow.com/questions/2230871/when-is-a-java-method-name-too-long/2230893#2230893) sums up how I think of this (without saying it is more right or wrong because it is on SO):

> A name in Java, or any other language, is too long when a shorter name exists that equally conveys the behavior of the method.

I think the most important with this recommendation is to discourage authors from creating their own abbreviations, which might be clear in the moment they write them, but not after that or for anyone else reading the code.